File: frontend/src/types/experiment.ts
// Experiment-related types

import type { UUID, Timestamp } from './base';
import type { MetricsTimeSeries, IslandState } from './metrics';

export const ExperimentStatus = {
  PENDING: 'pending',
  RUNNING: 'running',
  PAUSED: 'paused',
  COMPLETED: 'completed',
  FAILED: 'failed',
  STOPPED: 'stopped',
} as const;

export type ExperimentStatus = typeof ExperimentStatus[keyof typeof ExperimentStatus];

export interface LLMConfig {
  provider: 'lm_studio' | 'mock' | 'template_mock';
  model: string;
  base_url: string;
  temperature: number;
  max_tokens: number;
  timeout: number;
}

export interface SandboxConfig {
  provider: 'docker' | 'subprocess' | 'mock';
  max_workers: number;
  timeout: number;
  image: string;
  limits: {
    memory: string;
    cpu_cores: number;
  };
}

export interface FunSearchConfig {
  samples_per_prompt: number;
  num_islands: number;
  reset_period: number;
  cluster_sampling_temperature_init: number;
  cluster_sampling_temperature_period: number;
}

export interface ExecutionConfig {
  max_iterations: number;
  checkpoint_interval: number;
  log_level: 'DEBUG' | 'INFO' | 'WARNING' | 'ERROR';
}

export interface ExperimentConfig {
  llm: LLMConfig;
  sandbox: SandboxConfig;
  funsearch: FunSearchConfig;
  execution: ExecutionConfig;
}

export interface Experiment {
  id: UUID;
  project_id: UUID;
  name: string;
  status: ExperimentStatus;
  config: ExperimentConfig;
  started_at: Timestamp | null;
  completed_at: Timestamp | null;
  paused_at: Timestamp | null;
  iterations_completed: number;
  best_score: number | null;
  mlflow_run_id: string | null;
  task_id: string | null;
}

export interface ExperimentDetail extends Experiment {
  current_score: number | null;
  best_program: string | null;
  metrics: MetricsTimeSeries | null;
  island_states: IslandState[];
  error_message: string | null;
}

export interface CreateExperimentRequest {
  name: string;
  config: ExperimentConfig;
}

export interface ExperimentListResponse {
  experiments: Experiment[];
  total: number;
}

File: backend/models/experiment.py
"""Experiment database model."""

import enum
import uuid
from datetime import datetime
from typing import TYPE_CHECKING

from sqlalchemy import Column, DateTime, Enum, Float, ForeignKey, Integer, String, Text, UUID
from sqlalchemy.dialects.sqlite import JSON
from sqlalchemy.orm import relationship

from .base import Base

if TYPE_CHECKING:
    from .project import Project


class ExperimentStatus(str, enum.Enum):
    """Experiment status enum."""

    PENDING = "pending"
    RUNNING = "running"
    PAUSED = "paused"
    COMPLETED = "completed"
    FAILED = "failed"
    STOPPED = "stopped"


class Experiment(Base):
    """Experiment model."""

    __tablename__ = "experiments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    project_id = Column(
        UUID(as_uuid=True), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True
    )
    name = Column(String(255), nullable=False)
    status = Column(
        Enum(ExperimentStatus), nullable=False, default=ExperimentStatus.PENDING, index=True
    )
    config = Column(JSON, nullable=False)  # Store entire experiment config as JSON
    started_at = Column(DateTime, index=True)
    completed_at = Column(DateTime)
    paused_at = Column(DateTime)
    iterations_completed = Column(Integer, nullable=False, default=0)
    best_score = Column(Float)
    best_program = Column(Text)
    mlflow_run_id = Column(String(100), index=True)
    task_id = Column(String(100))  # Celery task ID
    error_message = Column(Text)
    created_at = Column(DateTime, nullable=False, default=datetime.utcnow)

    # Relationships
    project = relationship("Project", back_populates="experiments")

    def __repr__(self) -> str:
        return f"<Experiment(id={self.id}, name='{self.name}', status={self.status})>"
